### PR TITLE
osd: support schedule scrub between some time defined by users

### PR DESCRIPTION
**Pull request by XinzeChi**
Signed-off-by: Xinze Chi <xmdxcxz@gmail.com>

### CODE REVIEW

**Comment by liewegas** [comment]
Reviewed-by:

nice!

**Comment by athanatos** [comment]
This needs an explicit test in src/test.  Otherwise, a later change will kill this one.

**Comment by XinzeChi** [comment]
Add explicit  test based on @athanatos comments. Please review it. Thanks.

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -5834,8 +5834,40 @@ bool OSD::scrub_random_backoff()
   return false;
 }
 
+bool OSD::scrub_time_permit()
```
Comment (dzafman) [inline src/osd/OSD.cc]: If you had an optional arg that could over-ride the ceph_clock_now() call, the unites could test some realistic usage cases.

Comment (XinzeChi) [inline src/osd/OSD.cc]: @dzafman, can you give me some detail example?

Comment (liewegas) [inline src/osd/OSD.cc]: ceph_clock_now() actually already takes a cct now and can add in a time offset.  but we don't use it and not all callers pass the cct.  but a simpler test would be to just look at the current time and set a scrub range that either does or doesn't include now and verify that scrub_time_permit() returns the right value.

Comment (liewegas) [inline src/osd/OSD.cc]: alternatively, pass 'now' as an argument and then you can write a bunch of simple deterministic tests.  the caller is then just if (!scrub_time_permit(ceph_clock_now(NULL)) ...

Comment (XinzeChi) [inline src/osd/OSD.cc]: Looks Good, I will done.

Comment (dzafman) [inline src/osd/OSD.cc]: I was thinking of testing both osd_scrub_begin_hour < and > osd_scrub_end_hour.  So you would have I would test the following:

current time  osd_scrub_begin_hour    osd_scrub_end_hour    result
01:00

Comment (dzafman) [inline src/osd/OSD.cc]: I was thinking we should test both osd_scrub_begin_hour < and > osd_scrub_end_hour.  So I'd prefer we be able to inject a test time so we can do the following type of test cases.

current time  osd_scrub_begin_hour    osd_scrub_end_hour    result
01:00                  22                                        03                         true
20:00                  22                                         03                        false
04:00                 22                                          03                        false
02:00                 01                                          03                        true
00:30                 01                                          03                        false
08:00                 01                                          03                        false

Comment (dzafman) [inline src/osd/OSD.cc]: I'd add the time arg OSD::scrub_time_permit(utime_t now) and always pass ceph_clock_now(cct) from osd code.  The unit test would pass the time values to test with.

**Comment by XinzeChi** [comment]
Push another patch based on @liewegas  @dzafman comments.

**Comment by dzafman** [comment]
Reviewed-by David Zafman <dzafman@redhat.com>

**Pull request merged**